opencog/moses/scoring/precision_bscore.cc

```diff
@@ -450,388 +453,438 @@ behavioral_score precision_bscore::exact_selection(const scored_combo_tree_set&
+	pre_it head = tr.begin();
+	if (*head == id::logical_and)
+		++conj_n;
+	for (sib_it sib = head.begin(); sib != head.end(); ++sib)
+		if (*sib == id::logical_and)
+			++conj_n;
+	score_t conj_n_penalty = hardness * (-1.0 / (1.0 + conj_n));
+	bs.push_back(conj_n_penalty);
+
+	if (logger().is_fine_enabled())
+		logger().fine("precision = %f  conj_n=%u  conj_n penalty=%e",
+		              precision, conj_n, conj_n_penalty);
+
+	log_candidate_bscore(tr, bs);
+	return bs;
 }
 
-behavioral_score precision_conj_bscore::operator()(const combo_tree& tr) const
+behavioral_score precision_conj_bscore::operator()(const Handle &handle) const
 {
-    behavioral_score bs;
-
-    // compute active and sum of all active outputs
-    count_t active = 0.0;  // total weight of active outputs by tr
-    score_t sao = 0.0;     // sum of all active outputs (in the boolean case)
-    interpreter_visitor iv(tr);
-    auto interpret_tr = boost::apply_visitor(iv);
-    for (const CTable::value_type& vct : ctable) {
-        // vct.first = input vector
-        // vct.second = counter of outputs
-        if (interpret_tr(vct.first.get_variant()) == id::logical_true) {
-            double sumo = sum_outputs(vct.second);
-            count_t totalc = vct.second.total_count();
-            // For boolean tables, sao == sum of all true positives,
-            // and active == sum of true+false positives.
-            sao += sumo;
-            active += totalc;
-        }
-    }
-
-    // Compute normalized precision.  No hits means perfect precision :)
-    // Yes, zero hits is common, early on.
-    score_t precision = 1.0;
-    if (0 < active)
-        precision = sao / active;
-    bs.push_back(precision);
-
-    // Count the number of conjunctions (up to depth 2)
-    unsigned conj_n = 0;
-    typedef combo_tree::iterator pre_it;
-    typedef combo_tree::sibling_iterator sib_it;
-    pre_it head = tr.begin();
-    if (*head == id::logical_and)
-        ++conj_n;
-    for (sib_it sib = head.begin(); sib != head.end(); ++sib)
-        if (*sib == id::logical_and)
-            ++conj_n;
-    score_t conj_n_penalty = hardness * (-1.0 / (1.0 + conj_n));
-    bs.push_back(conj_n_penalty);
-
-    if (logger().is_fine_enabled())
-        logger().fine("precision = %f  conj_n=%u  conj_n penalty=%e",
-                     precision, conj_n, conj_n_penalty);
-
-    log_candidate_bscore(tr, bs);
-    return bs;
+	behavioral_score bs;
+
+	// compute active and sum of all active outputs
+	count_t active = 0.0;  // total weight of active outputs by tr
+	score_t sao = 0.0;     // sum of all active outputs (in the boolean case)
+
+	atomese::Interpreter interpreter(moses::compressed_value_key);
+	const ValuePtr result = interpreter(handle);
+	auto link_result = LinkValueCast(result)->value();
+	int i = 0;
+	for (const CTable::value_type &vct : ctable) {
+		// vct.first = input vector
+		// vct.second = counter of outputs
+		if (HandleCast(link_result.at(i))->get_type() == TRUE_LINK) {
```

Comment:
Better use `bool_value_to_bool` defined in

https://github.com/singnet/asmoses/blob/998557895ad4e39ea43a80fbef2f5eb75351640c/opencog/utils/valueUtils.h#L12-L15